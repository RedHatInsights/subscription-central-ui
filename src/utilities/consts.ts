export const supportLink = 'https://access.redhat.com/support';

export const subscriptionInventoryLink = window.location.pathname.replace('manifests', 'inventory');
