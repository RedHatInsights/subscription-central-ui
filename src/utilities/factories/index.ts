import manifestEntry from './manifestEntry';
import user from './user';

export default { manifestEntry, user };
