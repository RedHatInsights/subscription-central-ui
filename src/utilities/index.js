import { getBaseName, getPartialRouteFromPath } from './routeHelpers';

export { getBaseName, getPartialRouteFromPath };
