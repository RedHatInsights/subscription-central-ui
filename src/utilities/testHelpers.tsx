import React, { ReactNode } from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';

interface wrapperProps {
  children: ReactNode;
}

export const createQueryWrapper = () => {
  const queryClient = new QueryClient({ defaultOptions: { queries: { retry: false } } });
  const wrapper = ({ children }: wrapperProps) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  return wrapper;
};
