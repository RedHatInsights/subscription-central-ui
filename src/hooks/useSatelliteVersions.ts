import { useQuery } from 'react-query';
import { useToken } from '../utilities/platformServices';

interface SatelliteVersion {
  description: string;
  value: string;
}

const fetchSatelliteVersions = (jwtToken: Promise<string>) => async (): Promise<any> => {
  return fetch('/api/rhsm/v2/manifests/versions', {
    headers: { Authorization: `Bearer ${await jwtToken}` }
  }).then((response) => {
    if (response.status != 200) {
      throw new Error(`Failed to fetch satellite versions: ${response.statusText}`);
    }
    return response.json();
  });
};

const useSatelliteVersions = () => {
  const jwtToken = useToken();
  return useQuery('satelliteVersions', () => fetchSatelliteVersions(jwtToken)());
};

export { fetchSatelliteVersions, SatelliteVersion, useSatelliteVersions as default };
