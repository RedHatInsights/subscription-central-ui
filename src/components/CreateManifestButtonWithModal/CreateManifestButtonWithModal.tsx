import React, { FC, useState } from 'react';
import { Button } from '@patternfly/react-core';
import CreateManifestModal from '../CreateManifestModal';
import { User } from '../../hooks/useUser';

interface CreateManifestButtonWithModalProps {
  user: User;
}

const CreateManifestButtonWithModal: FC<CreateManifestButtonWithModalProps> = ({ user }) => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  const handleModalToggle = () => {
    setIsModalOpen(!isModalOpen);
  };

  return (
    <>
      <Button variant="primary" onClick={handleModalToggle} isDisabled={!user.canWriteManifests}>
        Create new manifest
      </Button>
      <CreateManifestModal handleModalToggle={handleModalToggle} isModalOpen={isModalOpen} />
    </>
  );
};

export default CreateManifestButtonWithModal;
