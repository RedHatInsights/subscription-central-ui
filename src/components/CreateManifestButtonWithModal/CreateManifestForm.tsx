import React, { useState, FC } from 'react';
import { Link } from 'react-router-dom';
import {
  Button,
  ActionGroup,
  Form,
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  FormGroup,
  TextInput,
  Title,
  Popover,
  FormSelect,
  FormSelectOption
} from '@patternfly/react-core';
import ExclamationCircleIcon from '@patternfly/react-icons/dist/js/icons/exclamation-circle-icon';
import CheckCircleIcon from '@patternfly/react-icons/dist/js/icons/exclamation-circle-icon';
import HelpIcon from '@patternfly/react-icons/dist/js/icons/help-icon';
import { useCreateSatelliteManifest } from '../../hooks/useCreateSatelliteManifest';
import { SatelliteVersion } from '../../hooks/useSatelliteVersions';
import { Processing } from '../emptyState';

interface CreateManifestFormProps {
  satelliteVersions: SatelliteVersion[];
  handleModalToggle: () => void;
  isModalOpen: boolean;
}

const CreateManifestForm: FC<CreateManifestFormProps> = ({
  satelliteVersions,
  handleModalToggle,
  isModalOpen
}) => {
  const [name, setName] = useState('');
  const [version, setVersion] = useState('Select type');
  const [nameValidated, setNameValidated] = useState('noval');
  const [versionValidated, setVersionValidated] = useState('noval');

  const { mutate, isLoading, isError, isSuccess } = useCreateSatelliteManifest();

  const handleNameChange = (value: string) => {
    if (name.length) {
      setNameValidated('noval');
    }
    setName(value);
  };

  const handleSelectChange = (value: string) => {
    setVersionValidated('noval');
    setVersion(value);
  };

  const clearErrors = () => {
    setNameValidated('noval');
    setVersionValidated('noval');
  };

  const validateForm = () => {
    clearErrors();
    let isValid = true;

    if (!name.length) {
      setNameValidated('error');
      isValid = false;
    }

    if (version === 'Select type') {
      setVersionValidated('error');
      isValid = false;
    }
    return isValid;
  };

  const handleFormSubmit = () => {
    const formIsValid = validateForm();
    if (!formIsValid) return;
    mutate({ name, version });
  };

  const resetForm = () => {
    setName('');
    setVersion('Select type');
    setNameValidated('noval');
    setVersionValidated('noval');
  };

  !isModalOpen && resetForm();

  const shouldShowForm = isLoading === false && isError === false && isSuccess === false;

  return (
    <>
      {shouldShowForm && (
        <>
          <Title headingLevel="h3" size="2xl">
            Create new manifest
          </Title>
          <p style={{ margin: '30px 0' }}>
            Creating a new manifest allows you to export subscriptions to your on-premise
            subscription management application.
          </p>
          <Form isWidthLimited>
            <FormGroup
              label="Name"
              helperTextInvalid="Please provide a name for your new manifest"
              validated={nameValidated}
              helperTextInvalidIcon={<ExclamationCircleIcon />}
              labelIcon={
                <Popover
                  bodyContent={
                    <div>
                      Provide a name that will help you associate this manifest with a specific
                      organization or on-premise subscription management application.
                    </div>
                  }
                >
                  <button
                    type="button"
                    aria-label="More info for name field"
                    onClick={(e) => e.preventDefault()}
                    aria-describedby="simple-form-name-01"
                    className="pf-c-form__group-label-help"
                  >
                    <HelpIcon noVerticalAlign />
                  </button>
                </Popover>
              }
              isRequired
              fieldId="simple-form-name-01"
            >
              <TextInput
                isRequired
                type="text"
                aria-describedby="simple-form-name-01-helper"
                value={name}
                placeholder="Name"
                onChange={handleNameChange}
              />
            </FormGroup>
            <FormGroup
              label="Type"
              helperTextInvalid="Please select a version for your new manifest"
              validated={versionValidated}
              labelIcon={
                <Popover
                  bodyContent={
                    <div>
                      Due to variation in supported features, it is important to match the type and
                      version of the subscription management application you are using.
                    </div>
                  }
                >
                  <button
                    type="button"
                    aria-label="More info for type field"
                    onClick={(e) => e.preventDefault()}
                    aria-describedby="simple-form-name-01"
                    className="pf-c-form__group-label-help"
                  >
                    <HelpIcon noVerticalAlign />
                  </button>
                </Popover>
              }
              isRequired
              fieldId="simple-form-name-01"
            >
              <FormSelect
                value={version}
                onChange={handleSelectChange}
                aria-label="FormSelect Input"
                validated={versionValidated}
              >
                <FormSelectOption
                  isDisabled={true}
                  key="select version"
                  value="Select type"
                  label="Select type"
                />
                {satelliteVersions?.map((satelliteVersion: SatelliteVersion) => {
                  const satelliteVersionName = satelliteVersion.value.replace('sat-', 'Satellite ');
                  return (
                    <FormSelectOption
                      isDisabled={false}
                      key={satelliteVersion.value}
                      value={satelliteVersion.value}
                      label={satelliteVersionName}
                    />
                  );
                })}
              </FormSelect>
            </FormGroup>
            <ActionGroup>
              <Button key="confirm" variant="primary" onClick={handleFormSubmit}>
                Save
              </Button>

              <Button key="cancel" variant="link" onClick={handleModalToggle}>
                Cancel
              </Button>
            </ActionGroup>
          </Form>
        </>
      )}
      {isLoading && (
        <EmptyState variant={EmptyStateVariant.small}>
          <Title headingLevel="h3">Creating manifest....</Title>
          <EmptyStateBody>
            <Processing />
          </EmptyStateBody>
        </EmptyState>
      )}
      {isSuccess && (
        <EmptyState variant={EmptyStateVariant.small}>
          <EmptyStateIcon icon={CheckCircleIcon} color="var(--pf-global--success-color--100)" />
          <Title headingLevel="h3">Success</Title>
          <EmptyStateBody>
            <h4>
              Your new manifest, <strong>{name}</strong>, has been successfully created.
            </h4>
            <Button
              key="close-window"
              variant="primary"
              onClick={handleModalToggle}
              style={{ marginTop: '20px' }}
            >
              Return to page
            </Button>
          </EmptyStateBody>
        </EmptyState>
      )}
      {isError && (
        <EmptyState variant={EmptyStateVariant.small}>
          <EmptyStateIcon
            icon={ExclamationCircleIcon}
            color="var(--pf-global--danger-color--100)"
          />
          <Title headingLevel="h3">Something went wrong</Title>
          <EmptyStateBody>
            <h4>
              Try refreshing the page. If the problem persists, contact your organization
              administrator or visit our <Link to="https://status.redhat.com/">status page</Link>{' '}
              for known outages.
            </h4>
            <Button
              key="close-window"
              variant="primary"
              onClick={handleModalToggle}
              style={{ marginTop: '20px' }}
            >
              Return to page
            </Button>
          </EmptyStateBody>
        </EmptyState>
      )}
    </>
  );
};

export default CreateManifestForm;
