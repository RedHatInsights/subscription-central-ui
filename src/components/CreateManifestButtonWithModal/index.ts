import CreateManifestButtonWithModal from './CreateManifestButtonWithModal';
export default CreateManifestButtonWithModal;
