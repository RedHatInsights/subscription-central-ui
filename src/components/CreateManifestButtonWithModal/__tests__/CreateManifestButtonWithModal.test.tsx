import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import CreateManifestButtonWithModal from '..';
import useSatelliteVersions, { SatelliteVersion } from '../../../hooks/useSatelliteVersions';
import { QueryClientProvider, QueryClient } from 'react-query';
import factories from '../../../utilities/factories';
import { createQueryWrapper } from '../../../utilities/testHelpers';
import '@testing-library/jest-dom/extend-expect';

jest.mock('../../../hooks/useUser');
jest.mock('../../../hooks/useSatelliteVersions');
const queryClient = new QueryClient();

it('renders the modal properly when the button is clicked', async () => {
  (useSatelliteVersions as jest.Mock).mockReturnValue({
    body: [] as SatelliteVersion[],
    isError: false,
    isLoading: false
  });

  const { getByText } = render(
    <QueryClientProvider client={queryClient}>
      <CreateManifestButtonWithModal user={factories.user.build({ canWriteManifests: true })} />
    </QueryClientProvider>
  );

  fireEvent.click(getByText('Create new manifest'));
  expect(
    getByText(
      'Creating a manifest allows you to export subscriptions to your on-premise subscription management application. Match the type and version of the subscription management application that you are using. All fields are required.'
    )
  ).toBeInTheDocument();
});

it('renders the Create manifest form with disabled button for user', async () => {
  (useSatelliteVersions as jest.Mock).mockReturnValue({
    body: [] as SatelliteVersion[],
    isLoading: false,
    data: []
  });
  const { getByText } = render(
    <QueryClientProvider client={queryClient}>
      <CreateManifestButtonWithModal user={factories.user.build({ canWriteManifests: false })} />
    </QueryClientProvider>
  );

  expect(getByText('Create new manifest').closest('button')).toHaveAttribute('disabled');
});
