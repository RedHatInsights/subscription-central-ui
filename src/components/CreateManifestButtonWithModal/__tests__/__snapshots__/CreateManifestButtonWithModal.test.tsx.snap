// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Manifest Button With Modal renders the modal properly when the button is clicked 1`] = `
<body
  class="pf-c-backdrop__open"
>
  <div
    aria-hidden="true"
  >
    <button
      aria-disabled="false"
      class="pf-c-button pf-m-primary"
      data-ouia-component-id="OUIA-Generated-Button-primary-1"
      data-ouia-component-type="PF4/Button"
      data-ouia-safe="true"
      type="button"
    >
      Create new manifest
    </button>
  </div>
  <div>
    <div
      class="pf-c-backdrop"
    >
      <div
        class="pf-l-bullseye"
      >
        <div
          aria-describedby="pf-modal-part-2"
          aria-label="Create satellite manifest"
          aria-labelledby="pf-modal-part-0"
          aria-modal="true"
          class="pf-c-modal-box pf-m-md"
          data-ouia-component-id="OUIA-Generated-Modal-medium-1"
          data-ouia-component-type="PF4/ModalContent"
          data-ouia-safe="true"
          id="pf-modal-part-0"
          role="dialog"
        >
          <button
            aria-disabled="false"
            aria-label="Close"
            class="pf-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Button-plain-1"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </button>
          
          <div
            class="pf-c-modal-box__body"
            id="pf-modal-part-2"
          >
            <h3
              class="pf-c-title pf-m-2xl"
              data-ouia-component-id="OUIA-Generated-Title-1"
              data-ouia-component-type="PF4/Title"
              data-ouia-safe="true"
            >
              Create new manifest
            </h3>
            <p
              style="margin: 30px 0px;"
            >
              Creating a new manifest allows you to export subscriptions to your on-premise subscription management application.
            </p>
            <form
              class="pf-c-form pf-m-limit-width"
              novalidate=""
            >
              <div
                class="pf-c-form__group"
              >
                <div
                  class="pf-c-form__group-label"
                >
                  <label
                    class="pf-c-form__label"
                    for="create-satellite-manifest-form-name"
                  >
                    <span
                      class="pf-c-form__label-text"
                    >
                      Name
                    </span>
                    <span
                      aria-hidden="true"
                      class="pf-c-form__label-required"
                    >
                       
                      *
                    </span>
                  </label>
                   
                  <span
                    aria-label="Provide a name that will help you associate this manifest with a specific organization or on-premise subscription management application."
                    tabindex="0"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      style="vertical-align: -0.125em;"
                      viewBox="0 0 1024 1024"
                      width="1em"
                    >
                      <path
                        d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
                      />
                    </svg>
                  </span>
                </div>
                <div
                  class="pf-c-form__group-control"
                >
                  <input
                    aria-invalid="false"
                    class="pf-c-form-control"
                    data-ouia-component-id="OUIA-Generated-TextInputBase-1"
                    data-ouia-component-type="PF4/TextInput"
                    data-ouia-safe="true"
                    id="create-satellite-manifest-form-name"
                    name="satelliteManifestName"
                    placeholder="Name"
                    required=""
                    type="text"
                    value=""
                  />
                  
                </div>
              </div>
              <div
                class="pf-c-form__group"
              >
                <div
                  class="pf-c-form__group-label"
                >
                  <label
                    class="pf-c-form__label"
                    for="create-satellite-manifest-form-type"
                  >
                    <span
                      class="pf-c-form__label-text"
                    >
                      Type
                    </span>
                    <span
                      aria-hidden="true"
                      class="pf-c-form__label-required"
                    >
                       
                      *
                    </span>
                  </label>
                   
                  <span
                    aria-label="Due to variation in supported features, it is important to match the type and version of the subscription management application you are using."
                    tabindex="0"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      style="vertical-align: -0.125em;"
                      viewBox="0 0 1024 1024"
                      width="1em"
                    >
                      <path
                        d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
                      />
                    </svg>
                  </span>
                </div>
                <div
                  class="pf-c-form__group-control"
                >
                  <select
                    aria-invalid="false"
                    aria-label="FormSelect Input"
                    class="pf-c-form-control"
                    data-ouia-component-id="OUIA-Generated-FormSelect-default-1"
                    data-ouia-component-type="PF4/FormSelect"
                    data-ouia-safe="true"
                    id="create-satellite-manifest-form-type"
                    name="satelliteManifestType"
                  >
                    <option
                      class=""
                      disabled=""
                      value=""
                    >
                      Select type
                    </option>
                  </select>
                  
                </div>
              </div>
              <div
                class="pf-c-form__group pf-m-action"
              >
                <div
                  class="pf-c-form__group-control"
                >
                  <div
                    class="pf-c-form__actions"
                  >
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-primary"
                      data-ouia-component-id="OUIA-Generated-Button-primary-2"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      id="save-manifest-button"
                      type="button"
                    >
                      Save
                    </button>
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-link"
                      data-ouia-component-id="OUIA-Generated-Button-link-1"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      id="cancel-create-manifest-button"
                      type="button"
                    >
                      Cancel
                    </button>
                  </div>
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
